runtime/lua/vim/treesitter/query.lua

```diff
@@ -249,7 +249,32 @@ local predicate_handlers = {
     end
 
     return false
-  end
+  end,
+
+  ["in-plain-string-set?"] = (function()
+    local sets = {}
+
+    return function(match, _, source, predicate)
+      local node = match[predicate[2]]
+      local start_row, start_col, end_row, end_col = node:range()
+      if start_row ~= end_row then
+        return false
+      end
+
+      local node_text = M.get_node_text(node, source)
+      local set_id = predicate[3]
```

Comment:
I guess the set id is a bit bug prone when directly exposed to the user. Can't get a id somehow directly from the query? I guess the verbatim query string is somewhere available in query info.

Reply:
```
{
  captures = <1>{ "variable", "keyword", "repeat", "conditional", "constant.macro", "include", "operator", "boolean", "punctuation.delimiter", "punctuation.bracket", "string", "constant.builtin", "number", "function", "function.macro", "p
roperty", "_parent", "label", "type", "constant", "_u", "comment", "parameter", "error", "field", "method", "namespace", "constructor", "variable.builtin", "exception", "keyword.operator", "attribute" },
  info = {
    captures = <table 1>,
    patterns = {
      [23] = { { "not-has-parent?", 17, "template_method", "function_declarator", "call_expression" } },
      [24] = { { "has-ancestor?", 16, "field_declaration" }, { "not-has-ancestor?", 16, "function_declarator" } },
      [30] = { { "match?", 20, "^[A-Z][A-Z0-9_]+$" } },
      [32] = { { "eq?", 21, "#undef" } },
      [38] = { { "match?", 25, "^_" } },
      [39] = { { "match?", 25, "^m_" } },
      [40] = { { "match?", 25, "_$" } },
      [45] = { { "has-parent?", 17, "template_method", "function_declarator", "call_expression" } },
      [49] = { { "match?", 19, "^[A-Z]" }, { "not-has-parent?", 19, "function_declarator" } },
      [51] = { { "match?", 19, "^[A-Z]" } },
      [52] = { { "match?", 20, "^[A-Z][A-Z_0-9]*$" } },
      [56] = { { "match?", 28, "^[A-Z]" } },
      [60] = { { "match?", 28, "^[A-Z]" } },
      [61] = { { "match?", 28, "^[A-Z]" } },
      [62] = { { "match?", 28, "^[A-Z]" } },
      [63] = { { "match?", 28, "^[A-Z]" } },
      [74] = { { "vim-match?", 32, "[?[.*]]?.*$" } }
    }
  },
```
I guess these ids stay constant for each occurrence of predicate

Reply:
What do you think about using `str()` to obtain a memory address of the predicate table? Too hack-ish?

Reply:
I guess the ID you need is just the parameter that you ignored. Can't say for sure. They ID is at least unique per query file so it needs to be combined with the query file.

I think you could even use the predicate table directly

```
local foo = {'da'}

foo[foo] = 'as'

print(vim.inspect(foo))
```
I think the table will not change and it will use the memory address of this table as hash.

Reply:
I decided to use the predicate table directly as a key. I also added an `else` branch to print "cache hit', and the code seems to just re-use the same table, so it is a stable key.